<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2. The `SpatialExperiment` class • EuroBioc2020SpatialExperimentWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2. The `SpatialExperiment` class">
<meta property="og:description" content="EuroBioc2020SpatialExperimentWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EuroBioc2020SpatialExperimentWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/HOWTO_BUILD_WORKSHOP.html">How To Build A Workshop Package</a>
    </li>
    <li>
      <a href="../articles/Introduction1.html">1. Introduction to Spatial Transcriptomics</a>
    </li>
    <li>
      <a href="../articles/SpaData1.html">3. The `SpaData` package</a>
    </li>
    <li>
      <a href="../articles/SpatialExperiment1.html">2. The `SpatialExperiment` class</a>
    </li>
    <li>
      <a href="../articles/workshop_example.html">An Example Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drighelli/EuroBioc2020_SpatialWorkshop">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SpatialExperiment1_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2. The <code>SpatialExperiment</code> class</h1>
                        <h4 class="author">Helena L. Crowell</h4>
            <address class="author_afil">
      Department of Molecular Life Sciences, University of Zurich, Zurich, Switzerland<br><a class="author_email" href="mailto:#"></a><a href="mailto:helena.crowell@uzh.ch" class="email">helena.crowell@uzh.ch</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/drighelli/EuroBioc2020_SpatialWorkshop/blob/master/vignettes/SpatialExperiment1.Rmd"><code>vignettes/SpatialExperiment1.Rmd</code></a></small>
      <div class="hidden name"><code>SpatialExperiment1.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SpatialExperiment</span><span class="op">)</span></code></pre></div>
<div id="reading-10x-visium-data" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-10x-visium-data" class="anchor"></a>Reading 10X Visium data</h1>
<p>The 10X Genomics’ <em>CellRanger</em> pipeline will process data using standard output file formats that are saved, for each sample, in a single directory <code>/&lt;sample&gt;/outs/</code> of the following structure:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="ex">sample</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">|</span>—<span class="ex">outs</span> </span>
<span id="cb2-3"><a href="#cb2-3"></a>··<span class="kw">|</span>—<span class="ex">raw/filtered_feature_bc_matrix.h5</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>··<span class="kw">|</span>—<span class="ex">raw/filtered_feature_bc_matrix</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>····<span class="kw">|</span>—<span class="ex">barcodes.tsv</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>····<span class="kw">|</span>—<span class="ex">features.tsv</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>····<span class="kw">|</span>—<span class="ex">matrix.mtx</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>··<span class="kw">|</span>—<span class="ex">spatial</span></span>
<span id="cb2-9"><a href="#cb2-9"></a>····<span class="kw">|</span>—<span class="ex">tissue_hires_image.png</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>····<span class="kw">|</span>—<span class="ex">tissue_lowres_image.png</span></span>
<span id="cb2-11"><a href="#cb2-11"></a>····<span class="kw">|</span>—<span class="ex">detected_tissue_image.jpg</span></span>
<span id="cb2-12"><a href="#cb2-12"></a>····<span class="kw">|</span>—<span class="ex">aligned_fiducials.jpg</span></span>
<span id="cb2-13"><a href="#cb2-13"></a>····<span class="kw">|</span>—<span class="ex">scalefactors_json.json</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>····<span class="kw">|</span>—<span class="ex">tissue_positions_list.csv</span></span></code></pre></div>
<p>The <code>SpatialExperiment</code> package provides an exemplary 10X Visium spatial gene expression data of two serial mouse brain sections (Sagittal-Posterior) available from the <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets">10X Genomics website</a>. These are located in the <code>extdata/10xVisium</code> directory:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"10xVisium"</span><span class="op">)</span>,
  package <span class="op">=</span> <span class="st">"SpatialExperiment"</span><span class="op">)</span>

<span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"section1"</span>, <span class="st">"section2"</span><span class="op">)</span>
<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span><span class="op">)</span></code></pre></div>
<p>We can load these data into a <code>SpatialExperiment</code> using the <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html">read10xVisium()</a></code> function, which will read in all relevant information, including the count data, spatial coordinates, scale factors, and images:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "raw_feature_bc_matrix.h5" "spatial"</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "scalefactors_json.json"    "tissue_lowres_image.png"  
## [3] "tissue_positions_list.csv"</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">ve</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html">read10xVisium</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>,
  images <span class="op">=</span> <span class="st">"lowres"</span>, <span class="co"># specify which image(s) to include</span>
  load <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>      <span class="co"># specify whether or not to load image(s)</span></code></pre></div>
<pre><code>## class: SpatialExperiment 
## dim: 32285 9984 
## metadata(2): Samples Samples
## assays(1): counts
## rownames(32285): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000095019 ENSMUSG00000095041
## rowData names(1): symbol
## colnames(9984): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...
##   TTGTTTGTATTACACG-1 TTGTTTGTGTAAATTC-1
## colData names(7): Barcode sample_id ... array_row array_col
## reducedDimNames(0):
## altExpNames(0):
## spatialCoordsNames(5) : x_coord y_coord in_tissue array_row array_col
## inTissue(1): 6710
## imgData(6): sample_id image_id ... height scaleFactor</code></pre>
</div>
<div id="the-spatialexperiment-class" class="section level1">
<h1 class="hasAnchor">
<a href="#the-spatialexperiment-class" class="anchor"></a>The <code>SpatialExperiment</code> class</h1>
<div id="spatial-data" class="section level2">
<h2 class="hasAnchor">
<a href="#spatial-data" class="anchor"></a>Spatial data</h2>
<p>Spatial data are stored as observation metadata (<code>colData</code>) and include:</p>
<ul>
<li>
<code>sample_id</code> specifying unique sample identifiers</li>
<li>
<code>in_tissue</code> indicating whether an observation was mapped to tissue</li>
<li>
<code>x/y_coord</code> storing spatial coordinates</li>
<li>
<code>array_row/col</code> giving the spots’ row/column coordinate in the array<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</li>
</ul>
<p>A <code>DataFrame</code> of spatially-related data can be accessed using the <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/spatialCoords.html">spatialCoords()</a></code> accessor:</p>
<pre><code>```r
head(spatialCoords(ve))
```

```
## DataFrame with 6 rows and 5 columns
##                      x_coord   y_coord in_tissue array_row array_col
##                    &lt;integer&gt; &lt;integer&gt; &lt;logical&gt; &lt;integer&gt; &lt;integer&gt;
## AAACAACGAATAGTTC-1      1419      2534     FALSE         0        16
## AAACAAGTATCTCCCA-1      7409      8455      TRUE        50       102
## AAACAATCTACTAGCA-1      1778      4393     FALSE         3        43
## AAACACCAATAACTGC-1      8487      2740      TRUE        59        19
## AAACAGAGCGACTCCT-1      3096      7905      TRUE        14        94
## AAACAGCTTTCAGAAG-1      6570      2052      TRUE        43         9
```</code></pre>
<p>Alternatively, we can access these data using <code>colData()</code> or, even simpler, the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> accessor:</p>
<pre><code>```r
# tabulate number of spots mapped to tissue
table(
in_tissue = ve$in_tissue,
sample_id = ve$sample_id)
```

```
##          sample_id
## in_tissue section1 section2
##     FALSE     1637     1637
##     TRUE      3355     3355
```</code></pre>
</div>
<div id="image-data" class="section level2">
<h2 class="hasAnchor">
<a href="#image-data" class="anchor"></a>Image data</h2>
<p>Image-related data are stored in the <code>int_metadata</code>’s <code>imgData</code> field as a <code>DataFrame</code> with the following columns:</p>
<ul>
<li>
<code>sample_id</code> and <code>image_id</code> specifying the image’s sample and image identifier</li>
<li>
<code>data</code>: a list of <code>SpatialImage</code>s containing the image’s <code>grob</code>, path and/or URL</li>
<li>
<code>width</code> and <code>height</code> giving the image’s dimension (in pixel)</li>
<li>
<code>scaleFactor</code> used to rescale spatial coordinates according to the image’s resolution</li>
</ul>
<p>We can retrieve these data using the <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData()</a></code> accessor:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 6 columns
##     sample_id    image_id   data     width    height scaleFactor
##   &lt;character&gt; &lt;character&gt; &lt;list&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;
## 1    section1      lowres              600       600   0.0516351
## 2    section2      lowres              600       600   0.0516351</code></pre>
</div>
</div>
<div id="the-spatialimage-class" class="section level1">
<h1 class="hasAnchor">
<a href="#the-spatialimage-class" class="anchor"></a>The <code>SpatialImage</code> class</h1>
<p>Images inside a <code>SpatialExperiment</code>’s <code>imgData</code> are stored as objects of class <code>SpatialImage</code>. These contain three slots that can accommodate any available information associated with an image:</p>
<pre><code>* `@grob`: NULL or an object class `rastergrob` from the `grid` package</code></pre>
<ul>
<li>
<code>@path</code>: NULL or a character strings specifying an image file name (.png, .jpg or .tif)</li>
<li>
<code>@url</code>: NULL or a character string specifying an URL from which to retrieve the image</li>
</ul>
<p>A list of <code>SpatialImage</code>s can be retrieved from the <code>imgData</code>’s <code>data</code> field using the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> accessor:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">$</span><span class="va">data</span></code></pre></div>
<pre><code>## [[1]]
## A SpatialImage with 2 source(s):
##       &gt; loaded
## grob: Av
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section1/spatial/tissue_lowres_ima
##       ge.png
##  url: NA
## 
## [[2]]
## A SpatialImage with 2 source(s):
##       &gt; loaded
## grob: Av
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section2/spatial/tissue_lowres_ima
##       ge.png
##  url: NA</code></pre>
<p>Data available in an object of class <code>SpatialImage</code> may be accessed via the <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgGrob()</a></code>, <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgPath()</a></code> and <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgUrl()</a></code> accessors:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">si</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">$</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgGrob</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span></code></pre></div>
<pre><code>## rastergrob[GRID.rastergrob.11]</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgPath</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "/usr/local/lib/R/site-library/SpatialExperiment/extdata/10xVisium/section1/spatial/tissue_lowres_image.png"</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgUrl</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span></code></pre></div>
<pre><code>## NULL</code></pre>
<p><code>grob</code>s can be used directly for plotting (e.g. using <code><a href="https://rdrr.io/r/grid/grid.draw.html">grid.draw()</a></code> or <code>ggplot2</code>’s <code><a href="https://ggplot2.tidyverse.org/reference/layer.html">layer()</a></code> and <code><a href="https://ggplot2.tidyverse.org/reference/annotation_custom.html">annotation_custom()</a></code>):</p>
<pre><code>```r
si &lt;- imgData(ve)$data[[1]]
grid.draw(imgGrob(si))
```

&lt;img src="/__w/EuroBioc2020_SpatialWorkshop/EuroBioc2020_SpatialWorkshop/docs/articles/SpatialExperiment1_files/figure-html/unnamed-chunk-10-1.png" width="700" /&gt;</code></pre>
<p><code>path</code> and <code>url</code> provide the option to store an image’s source at minimal storage cost. This is desirable when multiple images are to be stored (say, for many samples and of different resolutions), or when a <code>SpatialExperiment</code> is to be exported.</p>
</div>
<div id="methods-for-image-handling" class="section level1">
<h1 class="hasAnchor">
<a href="#methods-for-image-handling" class="anchor"></a>Methods for image handling</h1>
<p>The <code>SpatialExperiment</code> package provides various functions to handle which and how image data is stored in the object. These include:</p>
<ul>
<li>
<code>loadImg</code> to actively load (an) image(s) from a path or URL and store it as a <code>grob</code>
</li>
<li>
<code>unloadImg</code> to drop the <code>grob</code>, while retaining the source path and/or URL</li>
<li>
<code>addImg</code> to add a new image entry (as a path, URL, or <code>grob</code>)</li>
<li>
<code>removeImg</code> to drop an image entry entirely</li>
</ul>
<div id="loading-unloading-images" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-unloading-images" class="anchor"></a>Loading &amp; unloading images</h2>
<p><code><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">loadImg()</a></code> and <code>add/removeImg()</code> are flexible in the specification of the <code>sample/image_id</code> arguments. Specifically,</p>
<ul>
<li>
<code>TRUE</code> is equivalent to <em>all</em>, e.g. <code>sample_id = "&lt;sample&gt;", image_id = TRUE</code> will drop all images for a given sample.</li>
<li>
<code>NULL</code> defaults to the first entry available, e.g., <code>sample_id = "&lt;sample&gt;", image_id = NULL</code> will drop the first image for a given sample.</li>
</ul>
<p>For example, <code>sample_id,image_id = TRUE,TRUE</code> will specify all images; <code>NULL,NULL</code> corresponds to the first image entry in the <code>imgData</code>; <code>TRUE,NULL</code> equals the first image for all samples; and <code>NULL,TRUE</code> matches all images for the first sample.</p>
<p>In the example below, we unload all images, i.e., drop all <code>grob</code>s. As a result, <code>grob</code> slots will be set to <code>NULL</code>, and all <code>SpatialImage</code>s now say <code>&gt; not loaded</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ve</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">unloadImg</a></span><span class="op">(</span><span class="va">ve</span>, sample_id <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">$</span><span class="va">data</span></code></pre></div>
<pre><code>## [[1]]
## A SpatialImage with 1 source(s):
##       &gt; not loaded
## grob: NA
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section1/spatial/tissue_lowres_ima
##       ge.png
##  url: NA
## 
## [[2]]
## A SpatialImage with 1 source(s):
##       &gt; not loaded
## grob: NA
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section2/spatial/tissue_lowres_ima
##       ge.png
##  url: NA</code></pre>
<p>We can again reload a single or set of images using <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">loadImg()</a></code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ve</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">loadImg</a></span><span class="op">(</span><span class="va">ve</span>, sample_id <span class="op">=</span> <span class="st">"section2"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">$</span><span class="va">data</span></code></pre></div>
<pre><code>## [[1]]
## A SpatialImage with 1 source(s):
##       &gt; not loaded
## grob: NA
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section1/spatial/tissue_lowres_ima
##       ge.png
##  url: NA
## 
## [[2]]
## A SpatialImage with 2 source(s):
##       &gt; loaded
## grob: Av
## path: /usr/local/lib/R/site-library/SpatialExperiment/ex
##       tdata/10xVisium/section2/spatial/tissue_lowres_ima
##       ge.png
##  url: NA</code></pre>
</div>
<div id="adding-removing-images" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-removing-images" class="anchor"></a>Adding &amp; removing images</h2>
<p>Besides a path or URL to source the image from and a numeric scale factor, <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">addImg()</a></code> requires specification of the <code>sample_id</code> the new image belongs to, and an <code>image_id</code> that is not yet in use for that sample:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://i.redd.it/3pw5uah7xo041.jpg"</span>
<span class="va">ve</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">addImg</a></span><span class="op">(</span><span class="va">ve</span>,
  sample_id <span class="op">=</span> <span class="st">"section1"</span>, image_id <span class="op">=</span> <span class="st">"pomeranian"</span>,
  imageSource <span class="op">=</span> <span class="va">url</span>, scaleFactor <span class="op">=</span> <span class="cn">NA_real_</span>, load <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in sprintf(" 'image_id' and 'sample_id'", dQuote(c(image_id,
## sample_id))): one argument not used by format ' 'image_id' and 'sample_id''</code></pre>
<p>The above code chunk has added an new image entry in the input <code>SpatialExperiment</code>’s <code>imgData</code> field:</p>
<pre><code>```r
imgData(ve)
```

```
## DataFrame with 3 rows and 6 columns
##     sample_id    image_id   data     width    height scaleFactor
##   &lt;character&gt; &lt;character&gt; &lt;list&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;
## 1    section1      lowres              600       600   0.0516351
## 2    section2      lowres              600       600   0.0516351
## 3    section1  pomeranian             1200      1186          NA
```</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-methods.html">imgGrob</a></span><span class="op">(</span><span class="va">ve</span>,
               sample_id <span class="op">=</span> <span class="st">"section1"</span>,
               image_id <span class="op">=</span> <span class="st">"pomeranian"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grid/grid.draw.html">grid.draw</a></span><span class="op">(</span><span class="va">grb</span><span class="op">)</span></code></pre></div>
<p><img src="SpatialExperiment1_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We can remove specific images with <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html">removeImg()</a></code>:</p>
<pre><code>```r
ve &lt;- removeImg(ve,
            sample_id = "section1",
            image_id = "pomeranian")
imgData(ve)
```

```
## DataFrame with 2 rows and 6 columns
##     sample_id    image_id   data     width    height scaleFactor
##   &lt;character&gt; &lt;character&gt; &lt;list&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;
## 1    section1      lowres              600       600   0.0516351
## 2    section2      lowres              600       600   0.0516351
```</code></pre>
</div>
</div>
<div id="coldata-replacement" class="section level1">
<h1 class="hasAnchor">
<a href="#coldata-replacement" class="anchor"></a><code>colData</code> replacement</h1>
<p>While storing of <code>sample_id</code>s, the <code>in_tissue</code> indicator, and spatial <code>x/y_coords</code> inside the <code>SpatialExperiment</code>’s <code>colData</code> enables direct accessibility via the <code>colData</code> and <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> accessors, these fields are protected against arbitrary modification. This affects operations to the following effects:</p>
<p>Renaming is generally not permitted:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"x|y_coord"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"foo"</span></code></pre></div>
<pre><code>## Warning in .local(x, ..., value): cannot rename 'colData' fields 'x_coord',
## 'y_coord', 'in_tissue', 'array_row', 'array_col'</code></pre>
<p>Replacement of <code>sample_id</code>s is permitted provided that</p>
<ol style="list-style-type: lower-roman">
<li>the number of unique sample identifiers is retained</li>
<li>newly provided sample identifiers are a one-to-one mapping</li>
</ol>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ve</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span>, <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in .local(x, ..., value): Number of unique 'sample_id's is 2, but 3 were provided.
## Overwriting</code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ve</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">ve</span><span class="op">)</span>, <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in .local(x, ..., value): New 'sample_id's must map uniquely</code></pre>
<p>Valid replacement will be propagated to the <code>imgData</code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">ve</span>
<span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">ve</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span><span class="op">)</span>
<span class="va">tmp</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">imgData</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 6 columns
##     sample_id    image_id   data     width    height scaleFactor
##   &lt;character&gt; &lt;character&gt; &lt;list&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;
## 1     sample1      lowres              600       600   0.0516351
## 2     sample2      lowres              600       600   0.0516351</code></pre>
<p>The <code>x/y_coord</code> and <code>in_tissue</code> fields may be modified provided that the former is a logical vector, and the latter is a two- or three-column numeric matrix:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ve</span><span class="op">$</span><span class="va">x_coord</span> <span class="op">&lt;-</span> <span class="st">"x"</span>
<span class="va">ve</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&lt;-</span> <span class="st">"x"</span></code></pre></div>
<pre><code>## Warning in .local(x, ..., value): 'in_tissue' field in 'colData' should be
## 'logical'</code></pre>
<p><code>colData() &lt;- NULL</code> will retain only required fields, i.e. <code>sample_id</code>, <code>in_tissue</code> and <code>x/y_coord</code>:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Barcode"   "sample_id" "x_coord"   "y_coord"   "in_tissue" "array_row"
## [7] "array_col"</code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">ve</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">ve</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "sample_id" "x_coord"   "y_coord"   "in_tissue" "array_row" "array_col"</code></pre>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><code>array_row</code>s range from 0-77 (78 rows); <code>array_col</code>s are even in 0-126 for even rows, and odd in 1-127 for odd rows (64 columns), giving in <span class="math inline">\(78 \times 64 = 4,992\)</span> spots per sample.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Dario Righelli, Davide Risso.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
